examples/run/cuda/inst_nvtx.cpp

```diff
@@ -0,0 +1,88 @@
+/* Copyright (c) 2014, NVIDIA CORPORATION. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *  * Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *  * Neither the name of NVIDIA CORPORATION nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cxxabi.h>
+#include <dlfcn.h>
+#include <stdio.h>
+
+#include "nvtx3/nvToolsExt.h"
+#define UNUSED(x) (void)(x)
+const char *const default_name = "Unknown";
+
+const uint32_t colors[] = {0xff00ff00, 0xff0000ff, 0xffffff00, 0xffff00ff,
+                           0xff00ffff, 0xffff0000, 0xffffffff};
+const int num_colors = sizeof(colors) / sizeof(uint32_t);
+static int color_id = 0;
+
+extern "C" void __cyg_profile_func_enter(void *this_fn, void *call_site)
+    __attribute__((no_instrument_function));
+extern "C" void __cyg_profile_func_exit(void *this_fn, void *call_site)
+    __attribute__((no_instrument_function));
+
+void rangePush(const char *const name) __attribute__((no_instrument_function));
+void rangePop() __attribute__((no_instrument_function));
+
+extern "C" void __cyg_profile_func_enter(void *this_fn, void *call_site) {
+    (void)this_fn;
+    (void)call_site;
+    Dl_info this_fn_info;
+    if (dladdr(this_fn, &this_fn_info)) {
+        int status = 0;
+        rangePush(abi::__cxa_demangle(this_fn_info.dli_sname, 0, 0, &status));
+    } else {
+        rangePush(default_name);
+    }
```

Comment:
This is some prime platform specific crap right here. :frowning: If we do decide to add something like this to the project (which I'm still not convinced of), we'll need to protect it against being run on a platform where these things are not available.

Like:

```cmake
include( CheckIncludeFileCXX )
check_include_file_cxx( "cxxabi.h" TRACCC_HAVE_CXXABI_H )
if( HAVE_CXXABI_H )
   target_compile_definitions( traccc::cuda PRIVATE TRACCC_HAVE_CXXABI_H )
endif()
```

```c++
#ifdef TRACCC_HAVE_CXXABI_H
extern "C" {
#   include <cxxabi.h>
}
#endif // TRACCC_HAVE_CXX_ABI_H
...
#ifdef TRACCC_HAVE_CXX_ABI_H
int status = 0;
char* fname = abi::__cxa_demangle(this_fn_info.dli_sname, 0, 0, &status);
if (status != 0) {
   // Handle error somehow.
}
#else
const char* fname = this_fn_info.dli_sname;
#endif
...
#ifdef TRACCC_HAVE_CXX_ABI_H
free(fname);
#endif
```

I.e. yes, this expression also leaks some memory over here... :frowning: